include/dali/core/tensor_view.h

```diff
@@ -778,43 +778,95 @@ TensorListView<StorageBackend, DataType, output_ndim> sample_range(
+ * possibly with different element type.
+ * Non-contiguous input samples must not contribute to one output sample;
+ * merging contiguous samples and splitting is still possible.
  *
+ * @tparam U    new element type
  * @param list  original tensor list
  * @param shape the desired shape
  * @param check if true, exception is thrown when the list cannot be reshaped
  */
-template <int out_dim, typename Storage, typename T, int in_dim>
-TensorListView<Storage, T, out_dim> reshape(
-  const TensorListView<Storage, T, in_dim> &list,
-  TensorListShape<out_dim> shape,
-  bool check = false) {
-  if (check && shape.num_elements() != list.num_elements())
-    throw std::logic_error("Attempt to reshape a TensorListView to a different total volume");
-  assert(shape.num_elements() == list.num_elements());
-  if (list.is_contiguous()) {
-    return make_tensor_list<Storage, out_dim, T>(list.data[0], std::move(shape));
-  } else {
-    if (check && shape.num_samples() != list.num_samples())
-      throw std::logic_error("Attempt to reshape a non-contiguous TensorListView with a different "
-                             "number of samples");
-    assert(shape.num_samples() == list.num_samples());
-    for (int i = 0; i < list.num_samples(); i++) {
-      if (check && volume(list.tensor_shape_span(i)) != volume(shape.tensor_shape_span(i)))
-        throw std::logic_error("Attempt to change the volume of sample when reshaping a "
-                               "non-contiguous TensorListView");
-      assert(volume(list.tensor_shape_span(i)) == volume(shape.tensor_shape_span(i)));
+template <typename U, int out_dim, typename Storage, typename T, int in_dim>
+TensorListView<Storage, U, out_dim> reinterpret(
+      const TensorListView<Storage, T, in_dim> &list,
+      TensorListShape<out_dim> shape,
+      bool check = false) {
+  if (check && shape.num_elements() * sizeof(U) != list.num_elements() * sizeof(T))
+    throw std::logic_error("Attempt to reshape a TensorListView to a different total size");
+  assert(shape.num_elements() * sizeof(U) == list.num_elements() * sizeof(T));
+
+  int i = -1;  // start at index -1 - we'll increase it anyway
+  int M = list.num_samples();
+  int N = shape.num_samples();
+
+  if (!N)
+    return make_tensor_list<Storage, out_dim>(static_cast<U*>(nullptr), std::move(shape));
+
+  ptrdiff_t in_remaining = 0;
+  uintptr_t ptr = reinterpret_cast<uintptr_t>(list.data[0]);
+
+  TensorListView<Storage, U, out_dim> out_list;
+  out_list.shape = std::move(shape);
+  out_list.data.resize(N);
+
+  for (int o = 0; o < N; o++) {
+    ptrdiff_t out_remaining = volume(out_list.shape.tensor_shape_span(o)) * sizeof(U);
+    out_list.data[o] = reinterpret_cast<U *>(ptr);
+    bool first_chunk_in_sample = true;
+    while (out_remaining > 0) {
+      if (in_remaining > 0) {
+        ptrdiff_t to_add = out_remaining < in_remaining ? out_remaining : in_remaining;
```

Comment:
```suggestion
        ptrdiff_t to_add =std::min(out_remaining, in_remaining);
```

Reply:
That was (sort-of) deliberate `<algorithm>` is not included here and min/max are unfortunately there and not in `<utility>` :(